sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnarArray.java

```diff
@@ -46,7 +48,25 @@ public int numElements() {
 
   @Override
   public ArrayData copy() {
-    throw new UnsupportedOperationException();
+    DataType dt = data.dataType();
+
+    if (dt instanceof BooleanType) {
+      return UnsafeArrayData.fromPrimitiveArray(toBooleanArray());
+    } else if (dt instanceof ByteType) {
+      return UnsafeArrayData.fromPrimitiveArray(toByteArray());
+    } else if (dt instanceof ShortType) {
+      return UnsafeArrayData.fromPrimitiveArray(toShortArray());
+    } else if (dt instanceof IntegerType) {
+      return UnsafeArrayData.fromPrimitiveArray(toIntArray());
+    } else if (dt instanceof LongType) {
+      return UnsafeArrayData.fromPrimitiveArray(toLongArray());
+    } else if (dt instanceof FloatType) {
+      return UnsafeArrayData.fromPrimitiveArray(toFloatArray());
+    } else if (dt instanceof DoubleType) {
+      return UnsafeArrayData.fromPrimitiveArray(toDoubleArray());
+    } else {
+      throw new RuntimeException("Not implemented. " + dt);
```

Comment:
since we want to implement copy, let's complete it for all the types. We can create a `GenericArrayData` to copy data of non-primitive type, like what you did previously.

Reply:
Are you talking about the handling of  `StringType`, `DecimalType` and `TimestampType`?

Reply:
yes, we can exclude complex type(struct, array, etc.).

Reply:
Will it be correct for `StringType`?
```
    if (dt instanceof StringType) {
      GenericArrayData arrayData = new GenericArrayData(length);
      for (int i = 0; i < numElements(); i++) {
        arrayData.update(i, getUTF8String(i));
      }
      
      return arrayData;
    }
```

Reply:
@cloud-fan, `toArray` has two parameters

Reply:
maybe we can replace the body of method `clone` to `return new GenericArrayData(toArrayData(dt))`

Reply:
ah sorry, not `toArray`, we can use `toObjectArray`, which only need a data type.